### PR TITLE
[CALCITE-4988] ((A IS NOT NULL OR B) AND A IS NOT NULL) can't be simp…

### PR DESCRIPTION
**Pull request by NobiGo**
…lify to (A IS NOT NULL) When A is deterministic

### CODE REVIEW

**Comment by ILuffZhe** [comment]
Does this case cover the situation like "(A is not null and B is null) or A is not null"?

**Comment by NobiGo** [comment]
> Does this case cover the situation like "(A is not null and B is null) or A is not null"?

@ILuffZhe This can't cover this situation. The OR condition has a different way to pull up the predicate.

File: core/src/test/java/org/apache/calcite/rex/RexProgramTest.java [hunk 1]
```diff
@@ -2157,6 +2157,82 @@ private void checkExponentialCnf(int n) {
...
+    // "(A IS NOT NULL OR B) AND A IS NOT NULL" when A is not deterministic
+    // ==>
+    // "(A IS NOT NULL OR B) AND A IS NOT NULL"
+    SqlOperator ndc = getNoDeterministicOperator();
+    checkSimplifyUnchanged(
+        and(or(isNotNull(rexBuilder.makeCall(ndc)), gt(vInt(2), literal(2))),
+            isNotNull(rexBuilder.makeCall(ndc))));
+  }
+
+  @Test void testSimplifyIsNotNullWithNoDeterministic2() {
```
Comment (asolimando) [inline core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:2198]: I can't spot any difference w.r.t. `testSimplifyIsNotNullWithNoDeterministic`, is this a copy-paste error or am I missing something?

Comment (asolimando) [reply core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:2198]: Never mind, just saw the OR vs AND, my bad!

**Review by asolimando — APPROVED** [review]
LGTM!

**Pull request merged**